### PR TITLE
[DO NOT MERGE] Stop the user from losing changes to a schedule

### PR DESCRIPTION
**Pull request by Guntrisoft**
- if the user makes any changes and doesn't hit the save
  button then they will be alerted that they have changes
  that will be lost

### CODE REVIEW

File: app/assets/javascripts/calendars/guider-slot-picker.es6 [hunk 1]
```diff
@@ -20,13 +20,72 @@
 
       super.start(el);
 
-      this.addEvents();
-      this.generateJSON();
-      this.setupEvents();
+      this.saveWarningMessage = 'You have unsaved changes - Save, or undo the changes.';
```
Comment (benbarnett) [inline app/assets/javascripts/calendars/guider-slot-picker.es6]: Should this be in `TapBase` or `Calendar` now?

File: app/assets/javascripts/calendars/guider-slot-picker.es6 [hunk 2]
```diff
@@ -20,13 +20,72 @@
...
+      this.hiddenDataInput = $(this.$el.data('events'));
+      this.saveButton = $('.js-button-save');
+
+      this.addExistingEventsToCalendar();
```
Comment (benbarnett) [inline app/assets/javascripts/calendars/guider-slot-picker.es6]: This constructor is getting a little messy now 

Can we move these couple of lines into an `initCalendar()` method or something?

File: app/assets/javascripts/calendars/guider-slot-picker.es6 [hunk 3]
```diff
@@ -105,11 +151,27 @@
...
       }
 
-      dataElement.val(JSON.stringify(eventsOutput, null, 2));
+      return JSON.stringify(eventsOutput.sort(this.sortEvents.bind(this)));
     }
 
-    setupEvents() {
-      $(`#${this.$el.data('events-common')}`).find('button').on('click', this.handleEvent.bind(this));
+    calculateSortFingerprint(event) {
+      const minutes = (event.start_hour * 60) + event.start_minute;
```
Comment (benbarnett) [inline app/assets/javascripts/calendars/guider-slot-picker.es6]: 🙀

Do we have access to `moment` here?

File: app/assets/javascripts/calendars/guider-slot-picker.es6 [hunk 4]
```diff
@@ -122,20 +184,23 @@
       const events = $(event.currentTarget).data('events'),
       calendarView = this.$el.fullCalendar('getView'),
       calendarStartDate = calendarView.intervalStart,
-      calendarEndDate = calendarView.intervalEnd;
+      calendarEndDate = calendarView.intervalEnd.subtract(2, 'days'),
+      eventsToAdd = [];
 
       for (let currentDate = moment(calendarStartDate); currentDate < calendarEndDate; currentDate.add(1, 'days')) {
         for (let eventIndex in events) {
           event = events[eventIndex];
```
Comment (benbarnett) [inline app/assets/javascripts/calendars/guider-slot-picker.es6]: Have I misread this or are we overwriting the argument to the function here? We might have got round a missing lint warning here.

File: app/assets/javascripts/calendars/guider-slot-picker.es6 (hunk 2 above)
Comment (Guntrisoft) [inline app/assets/javascripts/calendars/guider-slot-picker.es6]: It's not strictly the constructor anymore (it's the start method), but I get your jist 👍 - will split out.

**Pull request closed without merging**